internal/sourcemap/metadata_fetcher.go

```diff
@@ -0,0 +1,320 @@
+		logger:           logp.NewLogger(logs.Sourcemap),
+		init:             make(chan struct{}),
+		invalidationChan: invalidationCh,
+	}
+
+	s.startBackgrounSync(ctx)
+
+	return s, invalidationCh
+}
+
+func (s *MetadataESFetcher) GetID(key Identifier) (*Identifier, bool) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	if _, ok := s.set[key]; ok {
+		return &key, ok
+	}
+
+	// path is missing from the metadata cache (and ES).
+	// Is it an alias ?
+	// Try to retrieve the sourcemap from the alias map
+	i, ok := s.alias[key]
+	return i, ok
+}
+
+func (s *MetadataESFetcher) Ready() <-chan struct{} {
+	return s.init
+}
+
+func (s *MetadataESFetcher) startBackgrounSync(parent context.Context) {
+	go func() {
+		s.logger.Debug("populating metadata cache")
+
+		// First run, populate cache
+		ctx, cancel := context.WithTimeout(parent, syncTimeout)
+		defer cancel()
+
+		if err := s.sync(ctx); err != nil {
+			s.logger.Errorf("failed to fetch sourcemaps metadata: %v", err)
+		} else {
+			// only close the init chan and mark the fetcher as ready if
+			// sync succeeded
+			close(s.init)
+		}
+
+		s.logger.Info("init routine completed")
+
+		go func() {
+			// TODO make this a config option ?
+			t := time.NewTicker(30 * time.Second)
+			defer t.Stop()
+
+			for {
+				select {
+				case <-t.C:
+					ctx, cancel := context.WithTimeout(parent, syncTimeout)
+
+					if err := s.sync(ctx); err != nil {
+						s.logger.Errorf("failed to sync sourcemaps metadata: %v", err)
+					}
+
+					cancel()
+				case <-parent.Done():
+					s.logger.Info("update routine done")
+					// close invalidation channel
+					close(s.invalidationChan)
+					return
+				}
+			}
+		}()
+	}()
+}
+
+func (s *MetadataESFetcher) sync(ctx context.Context) error {
+	sourcemaps := make(map[Identifier]string)
+
+	result, err := s.initialSearch(ctx, sourcemaps)
+	if err != nil {
+		return err
+	}
+
+	scrollID := result.ScrollID
+
+	if scrollID == "" {
+		s.update(ctx, sourcemaps)
+		return nil
+	}
+
+	for {
+		result, err = s.scrollsearch(ctx, scrollID, sourcemaps)
+		if err != nil {
+			return fmt.Errorf("failed scroll search: %w", err)
+		}
+
+		// From the docs: The initial search request and each subsequent scroll
+		// request each return a _scroll_id. While the _scroll_id may change between
+		// requests, it doesn't always change - in any case, only the most recently
+		// received _scroll_id should be used.
+		if result.ScrollID != "" {
+			scrollID = result.ScrollID
+		}
+
+		// Stop if there are no new updates
+		if len(result.Hits.Hits) == 0 {
+			break
+		}
+	}
+
+	s.update(ctx, sourcemaps)
+	return nil
+}
+
+func (s *MetadataESFetcher) update(ctx context.Context, sopurcemaps map[Identifier]string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	var invalidation []Identifier
+
+	for id, contentHash := range s.set {
+		if updatedHash, ok := sopurcemaps[id]; ok {
+			// already in the cache, remove from the updates.
+			delete(sopurcemaps, id)
```

Comment:
Since the ID is deleted from `sourcemaps` here, if `contenthash!=updatedHash`, the `s.set[id]=updatedHash` is missing.

Reply:
Good catch! Fixed :+1: